### PR TITLE
Generate DbContext Class in a Different Location

### PR DESCRIPTION
**Pull request by tonysneed**
Address #1836 by generating `DbContext` class to a different location.

For my first pass, I refactored `ReverseEngineeringScaffolder.Generate` to accept a `outputDbContextPath` parameter for specifying a different location for the `DbContext` class.

Next I'll look at how to refactor [OperationExecutor.ScaffoldContext](https://github.com/aspnet/EntityFrameworkCore/blob/dev/src/EFCore.Design/Design/OperationExecutor.cs#L417).

### CODE REVIEW

**Comment by tonysneed** [comment]
The CI build failed because of a timeout.  I ran the build and tests locally and they all passed, although it took more than an hour for them to complete.  Is there a more efficient way to run the tests?

**Comment by tonysneed** [comment]
@bricelam I implemented the update for the C# generator, but I'll probably need to do the same for VB and F#.  Can you point me in the right direction?

**Comment by tonysneed** [comment]
Currently there are a couple of failing tests, most likely because I'm doing something untoward.

One test, [DbContextScaffoldCommand_accepts_separate_context_output_path](https://github.com/aspnet/EntityFrameworkCore/pull/10356/files#diff-43caa941e5cdcb938d228b85a51c7491R35), fails with `LocalDB is not supported on this Platform`.  However, I've seen other tests using LocalDB without any problems.

The other test, [OperationExecutor_ScaffoldContext_generates_separate_context_output_path](https://github.com/aspnet/EntityFrameworkCore/pull/10356/files#diff-a765c5a543ff8870ccb999889c461a22R69), fails with a null reference exception on line [106](https://github.com/aspnet/EntityFrameworkCore/pull/10356/files#diff-a765c5a543ff8870ccb999889c461a22R106), probably because the `"ContextFile"` is not being generated.  This works for me locally, so there must be something about the build environment that is interfering with scaffolding the `DbContext` class.

Any help diagnosing these would be greatly appreciated! 😄

**Comment by bricelam** [comment]
> I implemented the update for the C# generator, but I'll probably need to do the same for VB and F#. Can you point me in the right direction?

I think the scaffolding code generators are responsible for way too much at the moment. I'm going to play around with the factoring a bit and try to move out the code that writes the files to disk. In Migrations, the flow looks something like this:

``` csharp
var scaffoldedMigration = migrationsScaffolder.ScaffoldMigration(); // Calls IMigrationsCodeGenerator
var files = migrationsScaffolder.Save(scaffoldedMigration);
```

This also helps with testing since you don't need to intercept the I/O using something like `IFileService`.

**Comment by bricelam** [comment]
P.S. We should also try and get some of this code out of the `.Internal` namespace for 2.1 so you can reliably depend on it.

**Comment by tonysneed** [comment]
Excellent - I’m  looking forward to the new flow and glad to play a part in it! And it will make Testing a lot easier too. :)

**Comment by tonysneed** [comment]
Waiting on #10459 to merge, then I'll refactor.

**Comment by smitpatel** [comment]
@tonysneed - If you can rebase on latest dev. 😄

**Comment by tonysneed** [comment]
I should have time for this later this week.

File: test/ef.Tests/CommandsTest.cs [hunk 1]
```diff
@@ -26,6 +28,61 @@ public void Short_names_are_unique()
...
+        [InlineData("FakeOutputDir", "FakeOutputDir")]
+        [InlineData("FakeOutputDir", "FakeDbContextOutputDir")]
+        public void DbContextScaffoldCommand_accepts_separate_context_output_path(string outputDir, string outputDbContextDir)
+        {
+            const string contextClassName = "FakeDbContextClassName";
+            var scaffoldCommand = GetScaffoldCommand();
+            var connectionString = new SqlConnectionStringBuilder
+            {
+                DataSource = @"(localdb)\MSSQLLocalDB",
+                InitialCatalog = "CommandConfiguration", // TODO: Ensure database exists
```
Comment (bricelam) [inline test/ef.Tests/CommandsTest.cs]: Don't bother with adding end-to-end tests yet. We have #9111 to figure out the best approach for this.

**Comment by tonysneed** [comment]
@bricelam You can review this, now that I've merged upstream commits and refactored.

I've added `outputDbContextDir` parameter to `IReverseEngineerScaffolder.Save`, which can be:

1. Null (defaults to `outputDir`)
2. Folder name under `projectPath`
3. Path relative to `projectPath`
4. Absolute path

Please let me know what you think.

File: src/EFCore.Design/Scaffolding/Internal/ReverseEngineerScaffolder.cs [hunk 2]
```diff
@@ -139,12 +150,26 @@ public class ReverseEngineerScaffolder : IReverseEngineerScaffolder
...
             CheckOutputFiles(scaffoldedModel, outputDir, overwriteFiles);
+            CheckOutputFiles(scaffoldedModel, outputDbContextDir, overwriteFiles);
 
             var files = new ReverseEngineerFiles();
             Directory.CreateDirectory(outputDir);
+            Directory.CreateDirectory(outputDbContextDir);
 
-            var contextPath = Path.Combine(outputDir, scaffoldedModel.ContextFile.Path);
+            var contextPath = Path.Combine(outputDbContextDir, scaffoldedModel.ContextFile.Path);
             File.WriteAllText(contextPath, scaffoldedModel.ContextFile.Code, Encoding.UTF8);
```
Comment (tonysneed) [inline src/EFCore.Design/Scaffolding/Internal/ReverseEngineerScaffolder.cs]: @bricelam There used to be an `IFileService` interface that abstracted the file system so you could mock it. Is there an alternative to that now?

**Comment by tonysneed** [comment]
I fixed the failing [test](https://github.com/aspnet/EntityFrameworkCore/pull/10356/commits/4977d9fd475c5150ee4770f8a168fd851951b0e0#diff-a765c5a543ff8870ccb999889c461a22R74) in commit 4977d9fd475c5150ee4770f8a168fd851951b0e0, and all the tests pass for me locally when running `build.cmd`.  So I'm not sure why the CI build would still be failing.

**Comment by tonysneed** [comment]
Wondering if someone  might be able to review this sometime soon?

**Comment by bricelam** [comment]
Sure, I think I'm done making breaks in this area. Can you rebase again?

**Comment by tonysneed** [comment]
@bricelam I rebased again and resolved conflicts.  I think the code it working properly at this point.  However, one of my tests (Microsoft.EntityFrameworkCore.Design OperationExecutor_ScaffoldContext_generates_separate_context_output_path) is getting an exception on `DatabaseOperations.ScaffoldContext` when it calls `services.GetRequiredService<IReverseEngineerScaffolder>()`:

```
Unable to resolve service for type 'Microsoft.EntityFrameworkCore.Storage.Converters.IValueConverterSelector' while attempting to activate 'Microsoft.EntityFrameworkCore.Storage.CoreTypeMapperDependencies'.
```

It looks like a dependency needs to be registered, but I'm not sure how to do that in the context of my unit test.  Would you be able to have a look?

**Comment by bricelam** [comment]
Changes look good. I suspect we just forgot to add the following to `DesignTimeServiceCollectionExtensions.AddEntityFrameworkDesignTimeServices()`.

``` csharp
.AddSingleton<IValueConverterSelector, ValueConverterSelector>()
```

**Comment by tonysneed** [comment]
Is this something I can fix? Or maybe open a separate issue?

**Comment by bricelam** [comment]
I can fix it as part of merging this.

**Comment by bricelam** [comment]
(We know we need #9111 to catch things like this.)

**Comment by tonysneed** [comment]
@bricelam Can you also remove the `Skip` parameter on the `[Theory]` attribute on my unit test?

**Comment by tonysneed** [comment]
Maybe this feature will make it into the next release?

**Comment by tonysneed** [comment]
Thanks! I think this tweak to the tooling will be appreciated by a lot of folks, especially those using repository patterns. /cc @lelong37

**Comment by bricelam** [comment]
@tonysneed Forgive me. I'm tweaking this slightly since #8434 will want to put the configuration classes next to the DbContext. There were also a couple of missing things (the corresponding PowerShell command and a resource). I removed some tests too 🐑 (sheepishly) since we need to figure out the best way to write functionals for RevEng (issue #9111), but I manually verified everything is working end to end.

Let me know if you have any concerns with my updates.

I also filed #10727 to clarify the semantics of this option.

**Comment by tonysneed** [comment]
@bricelam Looks great - thanks!

**Comment by tonysneed** [comment]
Glad to see this merged! 🎉

**Comment by weitzhandler** [comment]
Thanks @tonysneed!

**Pull request merged**